src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java

```diff
@@ -16,6 +20,51 @@ public void setup() {
 
     @Test
     public void testFindProductsNoResults() {
-        assertTrue(productDao.findProducts().isEmpty());
+        assertFalse(productDao.findProducts().isEmpty());
+    }
+
+    @Test
+    public void testFindProductWithZeroStock() throws ProductNotFoundException {
+        Currency usd = Currency.getInstance("USD");
+        Product product = new Product("test-product", "Samsung Galaxy S", new BigDecimal(100), usd, -1, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Samsung/Samsung%20Galaxy%20S.jpg");
+        productDao.save(product);
+        assertTrue(product.getId() > 0);
+        Product result = productDao.getProduct(product.getId());
+        assertNotNull(result);
+        assertEquals(result.getCode(), "test-product");
+    }
+
+    @Test
+    public void testSaveProductWithExistingId() throws ProductNotFoundException {
+        Currency usd = Currency.getInstance("USD");
+        Product product = new Product("test-product", "Samsung Galaxy S", new BigDecimal(100), usd, -1, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Samsung/Samsung%20Galaxy%20S.jpg");
+        productDao.save(product);
+        assertTrue(product.getId() > 0);
+        Product product1 = new Product(product.getId(), "new-test-product", "Samsung Galaxy S", new BigDecimal(100), usd, -1, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Samsung/Samsung%20Galaxy%20S.jpg");
+        productDao.save(product1);
+        Product result = productDao.getProduct(product.getId());
+        assertNotNull(result);
+        assertEquals(result.getCode(), "new-test-product");
```

Comment:
in all the assert-s the order is important - the first object is "expectedObject", the second is "actualObect"